cupy/random/_bit_generator.pyx

```diff
@@ -0,0 +1,184 @@
+        return sizeof(curandState)
+
+class MRG32k3a(_cuRANDGenerator):
+    generator = CURAND_MRG32k3a
+
+    def _type_size(self):
+        return sizeof(curandStateMRG32k3a)
+
+
+class Philox4x3210(_cuRANDGenerator):
+    generator = CURAND_PHILOX_4x32_10
+
+    def _type_size(self):
+        return sizeof(curandStatePhilox4_32_10_t)
+
+
+class Generator:
+    def __init__(self, bit_generator):
+        self._bit_generator = bit_generator 
+
+    def integers(self, low, high, size, dtype=numpy.int32, endpoint=False):
+        cdef ndarray y
+
+        diff = high-low
+        if not endpoint:
+           diff -= 1
+
+        cdef uint64_t mask = (1 << diff.bit_length()) - 1
+        # TODO adjust dtype
+        if diff <= _UINT32_MAX:
+            dtype = numpy.uint32
+        elif diff <= _UINT64_MAX:
+            dtype = numpy.uint64
+        else:
+            raise ValueError(
+                'high - low must be within uint64 range (actual: {})'.format(diff))
+
+        y = ndarray(size if size is not None else (), dtype)
+
+        if dtype is numpy.uint32:
+           self._launch_distribution_kernel(interval_32, y, diff, mask)
+        else:
+           self._launch_distribution_kernel(interval_64, y, diff, mask)
+        return low + y
+
+    def beta(self, a, b, size=None, dtype=float):
+        """Returns an array of samples drawn from the beta distribution.
+
+        .. seealso::
+            :func:`cupy.random.beta` for full documentation,
+            :meth:`numpy.random.RandomState.beta
+            <numpy.random.mtrand.RandomState.beta>`
+        """
+        cdef ndarray y
+        # cdef uint64_t state = <uint64_t>self._bit_generator.state()
+        y = ndarray(size if size is not None else (), dtype)
+        self._launch_distribution_kernel(beta, y, a, b)
+        return y
+
+    def standard_exponential(self, size=None, dtype=numpy.float64, method='inv', out=None):
+        cdef ndarray y
+
+        if method == 'zig':
+            raise NotImplementedError('Ziggurat method is not supported')
+                 
+        y = ndarray(size if size is not None else (), dtype)
+        self._launch_distribution_kernel(exponential, y)
+        if out is not None:
+            out[...] = y
+            y = out
+        return y
+
+    def _launch_distribution_kernel(self, func, out, *args):
+        # The generator might only have state for a few number of threads,
+        # what we do is to split the array filling in several chunks that are
+        # generated sequentially using the same state
+        cdef intptr_t strm = stream.get_current_stream_ptr()
+        state_ptr = self._bit_generator.state()
+        cdef state = <intptr_t>state_ptr
+        cdef y_ptr = <intptr_t>out.data.ptr
+        cdef ssize_t size = out.size
+        cdef int generator = self._bit_generator.generator
+
+        cdef bsize = self._bit_generator.state_size()
+        if bsize == 0:
+            func(state, y_ptr, out.size, strm, *args)
```

Comment:
Why don't we need `generator` as the first argument here?

Reply:
LOL

Reply:
I am wondering why it worked....